### ISSUE TITLE
Error spam in 2.1.0: Git executable not found, cannot extract git metadata

### ISSUE THREAD

**Issue by drichardson** [DataDog/dd-trace-py#7257]
### Summary of problem

We updated to 2.1.0 yesterday and are now getting error spam "Git executable not found, cannot extract git metadata" in production (i.e. this is not a CI log, it's being logged in production).

We don't have `git` installed in production nor do we want `git` installed in production.

I've also filed a ticket with datadog support here https://help.datadoghq.com/hc/en-us/requests/1391147

### Which version of dd-trace-py are you using?

2.1.0

### Which version of pip are you using?

23.1.2

### Which libraries and their versions are you using?

Leaving this out intentionally.

### How can we reproduce your problem?

Not sure, but guessing if you run ddtrace-run like you would in production without `git` being installed.

### What is the result that you get?

This is being spammed in production as an error log starting in 2.1.0:
```
Git executable not found, cannot extract git metadata
```

### What is the result that you expected?

No error log spam.

**Comment by heavenshell** [comment]
Same error raised at 2.1.0

**Comment by Tenzer** [comment]
We are seeing the same errors. It looks like there's an environment variable that can be set to disable this: https://github.com/DataDog/dd-trace-py/blob/d2ce0d063c432617bd45a51f974c2e319ab8e987/ddtrace/internal/gitmetadata.py#L19-L20

**Comment by doctornkz-intelas** [comment]
The same issue, and I'd suggest disabling this feature by default.

**Comment by romainkomorndatadog** [comment]
@drichardson , thanks for the report and the support task. Seems like these are coming from the CIVisibility recorder (and as you said, that shouldn't be happening in a non-CI context). I'll take a look at it.

**Comment by romainkomorndatadog** [comment]
One potential suspect is our [new unittest integration](https://github.com/DataDog/dd-trace-py/blob/2.x/ddtrace/contrib/unittest/patch.py#L295), but I can't reproduce in a simple `ddtrace-run` test case without explicitly doing an `import unittest`. In that case, `ddtrace-run` with `DD_PATCH_MODULES=unittest:false` set should prevent the log span.

@drichardson , is there any chance your app is importing the `unittest` module at some point? Would you be able to try the `DD_PATCH_MODULES` env var to validate whether or not that silences the messages?

**Comment by drichardson** [comment]
yeah we have one file that does in a flask cli we use for testing, but it probably get's loaded in prod ;-;

**Comment by drichardson** [comment]
This is at the top of said file

```
import logging
import time
from datetime import datetime
from unittest.mock import patch
```

**Comment by drichardson** [comment]
Trying out the DD_PATCH_MODULES thing now, will get back to you.

**Comment by drichardson** [comment]
Confirmed `DD_PATCH_MODULES=unittest:false` git rid of the log spam.

**Comment by romainkomorndatadog** [comment]
@heavenshell , @doctornkz-intelas , @Tenzer , are you in similar scenarios where the `unittest` module's imported in your prod envs?

**Comment by Tenzer** [comment]
I just checked, and yes, it turns out we are. I'll be switching to the `DD_PATCH_MODULES` environment variable as well.

**Comment by Tenzer** [comment]
In our case we are using `unittest.mock.patch` on a third-party library, to fix a problem in their code without having to fork the library.

**Comment by romainkomorndatadog** [comment]
I'm going to ask @ericlaz to take a look at this when he gets back later this week.

**Comment by romainkomorndatadog** [comment]
The "failed to shutdown tracer" exception sounds potentially related to #7273 / #7214 .

The fact that `DD_PATCH_MODULES=unittest:false` worked for you, @jerr0328 , definitely makes it sound like something else is importing `unittest`. I can't think of another reason why not patching the `unittest` module would make the spam go away.

**Comment by adilnaimi** [comment]
We faced a similar issue with version 2.1.1, although we don't have unittest. Our app has two folders, 'src' and 'tests.' During the image build process in ECS, only the 'src' folder is copied to the container image. We attempted to resolve the issue by setting `DD_PATCH_MODULES=unittest: false,` but it didn't help. We had to downgrade to 'ddtrace=2.0.3'. The app started OK, and the error disappeared. 
![CleanShot 2023-10-19 at 16 46 23](https://github.com/DataDog/dd-trace-py/assets/1929963/922deb03-da25-4644-839a-61e0d1ddec2b)

**Comment by doctornkz-intelas** [comment]
With 2.1.1 version it's getting even worse:
```
2023-10-20 07:33:03,742 ddtrace.internal.ci_visibility.recorder WARNING CODEOWNERS file is not available
2023-10-20 07:33:03,927 ddtrace.ext.git ERROR Git executable not found, cannot extract git metadata.
Git executable not found, cannot extract git metadata.
Git executable not found, cannot extract git metadata.
2023-10-20 07:33:03,938 ddtrace.ext.git ERROR Git executable not found, cannot extract git metadata.
Git executable not found, cannot extract git metadata.
2023-10-20 07:33:03,944 ddtrace.ext.ci ERROR Git executable not found, cannot extract git metadata.
2023-10-20 07:33:03,956 ddtrace.internal.ci_visibility.recorder WARNING CODEOWNERS file is not available
CODEOWNERS file is not available
```
Also, I see exceptions the same as @jerr0328 :
![Screenshot 2023-10-20 at 08 44 14](https://github.com/DataDog/dd-trace-py/assets/116017107/bc74a1a6-a74a-48f8-a2b3-ebf715ac5030)

I tried to use `DD_TRACE_GIT_METADATA_ENABLE=false` but I figured out that I have too many places to fix - so I locked `ddtrace` version to 2.0.1. 

I have to mention that we are not using `unittest` import in our code, but it might be imported somewhere in Django framework. I don't like the idea of disabling `ddtrace` mechanic in our application here and there just to make it work, it's not predictable.

I believe you will find a better solution.

**Comment by romainkomorndatadog** [comment]
@doctornkz-intelas , I agree that having to turn off functionality to avoid the log messages isn't a good solution. That's why we've got https://github.com/DataDog/dd-trace-py/pull/7292 which should address this by only enabling the `CIVisibility` recorder when `unittest` is imported if one of the functions that actually run tests is used. That should resolve this without need to set `DD_PATCH_MODULES` or `DD_TRACE_GIT_METADATA` in the environment.

It should ship in versions `2.2` and `2.1.2`.

It's already merged in our `2.x` branch, so you should be able to ```pip install git+https://github.com/DataDog/dd-trace-py.git@02b0eeb23388eda24c5c140ea5cbc01bb89f544f``` and verify the errors aren't being logged anymore.

**Comment by doctornkz-intelas** [comment]
Thank you, @romainkomorndatadog ! It looks like the way to go.

**Comment by adilnaimi** [comment]
> @adilnaimi , just to double check: the space you have in `DD_PATCH_MODULES=unittest: false` is just in this issue comment, not in your shell command? I find it odd that downgrading `ddtrace` would stop the warnings, but `DD_PATCH_MODULES=unittest:false` wouldn't. I haven't been able to reproduce the error logging in any way that doesn't involve importing `unittest`.
> 
> Could you double check that none of your dependencies are importing `unittest` (eg: by looking at `sys.modules`)?

@romainkomorndatadog - the space I have in `DD_PATCH_MODULES=unittest: false` is just in my Github comment.
I confirm none of your dependencies are importing `unittest`.

**Comment by romainkomorndatadog** [comment]
@adilnaimi , is there any chance you could try the fix we submitted in #7292 by using this in place of `pip install ddtrace` in a test run?

```
pip install git+https://github.com/DataDog/dd-trace-py.git@02b0eeb23388eda24c5c140ea5cbc01bb89f544f
```

I'm mostly curious to see if this does not fix the issue since it means there's some other path that ends up enabling the `CIVisibility` service (which I haven't found yet).

**Comment by adilnaimi** [comment]
@romainkomorndatadog -- I confirm using the version https://github.com/DataDog/dd-trace-py/pull/7292 fixes the issue:

### v2.1.1 
```
root@4cf25c2c2506:/app# gunicorn \
        --bind 0.0.0.0:8000 \
        --workers 2 \
        --worker-class src.asgi.worker.UvicornWorker \
        --log-file - \
        --access-logfile - \
        --log-level info \
        --keep-alive=60 \
        --backlog="${GUNICORN_BACKLOG:-1000}" \
        src.asgi:application
2023-10-24T11:15:05.805123Z [error    ] Git executable not found, cannot extract git metadata. [ddtrace.ext.git]
2023-10-24T11:15:05.805183Z [error    ] Git executable not found, cannot extract git metadata. [ddtrace.ext.git]
2023-10-24T11:15:05.805707Z [error    ] Git executable not found, cannot extract git metadata. [ddtrace.ext.git]
2023-10-24T11:15:05.805751Z [error    ] Git executable not found, cannot extract git metadata. [ddtrace.ext.git]
2023-10-24T11:15:05.806232Z [error    ] Git executable not found, cannot extract git metadata. [ddtrace.ext.ci]
2023-10-24T11:15:05.806271Z [error    ] Git executable not found, cannot extract git metadata. [ddtrace.ext.ci]
2023-10-24T11:15:05.849794Z [info     ] Starting gunicorn 21.2.0       [gunicorn.error]
2023-10-24T11:15:05.850249Z [info     ] Listening at: http://0.0.0.0:8000 (7) [gunicorn.error]
2023-10-24T11:15:05.850295Z [info     ] Using worker: src.asgi.worker.UvicornWorker [gunicorn.error]
2023-10-24T11:15:05.856546Z [info     ] Booting worker with pid: 16    [gunicorn.error]
2023-10-24T11:15:05.962159Z [info     ] Booting worker with pid: 19    [gunicorn.error]
```

### Version from the fix https://github.com/DataDog/dd-trace-py/pull/7292
```
root@4cf25c2c2506:/app# pip freeze | grep ddtrace
ddtrace @ git+https://github.com/DataDog/dd-trace-py.git@02b0eeb23388eda24c5c140ea5cbc01bb89f544f
root@4cf25c2c2506:/app# gunicorn \
        --bind 0.0.0.0:8000 \
        --workers 2 \
        --worker-class src.asgi.worker.UvicornWorker \
        --log-file - \
        --access-logfile - \
        --log-level info \
        --keep-alive=60 \
        --backlog="${GUNICORN_BACKLOG:-1000}" \
        src.asgi:application
2023-10-24T11:36:33.603836Z [info     ] Starting gunicorn 21.2.0       [gunicorn.error]
2023-10-24T11:36:33.604304Z [info     ] Listening at: http://0.0.0.0:8000 (4075) [gunicorn.error]
2023-10-24T11:36:33.604355Z [info     ] Using worker: src.asgi.worker.UvicornWorker [gunicorn.error]
2023-10-24T11:36:33.610415Z [info     ] Booting worker with pid: 4081  [gunicorn.error]
2023-10-24T11:36:33.681193Z [info     ] Booting worker with pid: 4084  [gunicorn.error]
```

**Comment by romainkomorndatadog** [comment]
@adilnaimi great, thanks!

**Comment by romainkomorndatadog** [comment]
The fix in #7292 shipped in [ddtrace==2.1.2](https://github.com/DataDog/dd-trace-py/releases/tag/v2.1.2). I'm going to close out this issue since it appears to fix all the reports.
